Enable the Nextcloud Sync Service to mark a storage as unhealthy

mereghost opened the pull request:
##### Related Work Package: [OP#50914](https://community.openproject.org/projects/openproject/work_packages/50914) 

## What?

The sync jobs run on the background, frequently failing without knowledge of the admins. This PR aims to lay the groundwork for the SyncService/CronJob pair to mark the storage as unhealthy.

## How?

I added a new `jsonb` field to the storage, containing the attributes that we would need to notify an admin that something is wrong with their integration.

The names right now are... dumb, I'm all ears for better suggestions.

Moreover I've made changes so that the entire SyncService works with service results, which will make it easier to return the necessary info.

mereghost committed:
Initial mess

mereghost committed:
small rearranging of the tests

mereghost committed:
Move the migration to storages module

Kharonus commented on modules/storages/app/common/storages/peripherals/storage_interaction/nextcloud/add_user_to_group_command.rb:
```diff
@@ -53,33 +53,36 @@ def call(user:, group: @group)
           )
       )
 
+      error_data = Storages::StorageErrorData.new(source: self.class, payload: response)
```
🟢 Nice, thanks!

Kharonus commented on modules/storages/app/models/storages/nextcloud_storage.rb:
```diff
@@ -41,20 +41,7 @@ class NextcloudStorage < Storage
     store_attribute :provider_fields, :group, :string
     store_attribute :provider_fields, :group_folder, :string
 
-    def self.sync_all_group_folders
```
❓  can you shortly elaborate, why this class method is no longer needed? I fail to see the connection to the scope of the PR, yet.

Kharonus replied:
Ah, is it, that you basically moved the code into the cron job?

mereghost replied:
Yep. This was moved to bring it closer to what triggers the behaviour.

Kharonus commented on modules/storages/app/services/storages/group_folder_properties_sync_service.rb:
```diff
@@ -46,40 +46,40 @@ class GroupFolderPropertiesSyncService
     end
 
     def call
-      prepare_remote_folders and apply_permissions_to_folders
```
🟡 why can't we bind both calls together? so that a failure in `prepare_remote_folder` would already return failure service result? We did this with the `>>` operator, didn't we?

The new lines contain the same error handling twice.

mereghost replied:
`bind` is sure an option. I've changed it a bit to a syntax that may look clearer to the casual rubyist.

Kharonus commented on modules/storages/app/services/storages/group_folder_properties_sync_service.rb:
```diff
@@ -90,35 +90,27 @@ def apply_permissions_to_folders
-                                         message: error.log_message,
-                                         data: { status: error.data.code, body: error.data.body } }.to_json)
+        return format_and_log_error(error, group: @storage.group)
+        # ServiceResult.failure(errors: error)
```
🔴 commented line

Kharonus commented on modules/storages/app/workers/storages/manage_nextcloud_integration_cron_job.rb:
```diff
@@ -26,14 +28,25 @@
+          result = GroupFolderPropertiesSyncService.call(storage)
+          result.error_and do |error|
+            # Sample code
+            storage.mark_as_unhealthy(reason: error.code)
```
🟡 a bit confused: how does a storage will get healthy again? I see no branch where it is marked as healthy.

mereghost committed:
Some cleanup after talking with Eric

ba1ash commented on modules/storages/app/services/storages/group_folder_properties_sync_service.rb:
```diff
@@ -46,40 +46,37 @@ class GroupFolderPropertiesSyncService
-        raise error_msg if error.code == :error
-
-        return OpenProject.logger.warn error_msg
+        # This hides the folder, but it will be in the logs
```
I cannot understand the comment here.
The return of failure directly will not hide the folder, right. I suggest  to be more explicit here:
1. clarify which folder: `the root group folder`
2. have a reference how returning a failure result leads to hiding.

mereghost replied:
~~The comment is wrong XD It should say, "don't hide the folder" 🤦🏼 ~~

Copied the comment when replacing lines 🤦🏼 ^2

mereghost committed:
Remove stupid comments

ba1ash committed:
Merge branch 'dev' into impl/add-health-info-to-storages

ba1ash committed:
Merge branch 'dev' into impl/add-health-info-to-storages

ba1ash committed:
Move sotrage models specs to storages/.

Because Rubocop forces this convention.

ba1ash committed:
Do not use Storages::Storage model in migrations.

It breaks the migration process if we define store_attribute and
try to use model class without store_attribute column in the db.
NoMethodError: undefined method `accessor' for #<ActiveModel::Type::Value:0x00007f7aa00163b0 @precision=nil, @scale=nil, @limit=nil> (NoMethodError)

        @store_accessor = subtype.accessor
                                 ^^^^^^^^^
/usr/local/bundle/gems/store_attribute-1.1.1/lib/store_attribute/active_record/type/typed_store.rb:33:in `initialize'
/usr/local/bundle/gems/store_attribute-1.1.1/lib/store_attribute/active_record/type/typed_store.rb:25:in `new'
/usr/local/bundle/gems/store_attribute-1.1.1/lib/store_attribute/active_record/type/typed_store.rb:25:in `create_from_type'
/usr/local/bundle/gems/store_attribute-1.1.1/lib/store_attribute/active_record/store.rb:188:in `block in _define_store_attribute'

ba1ash committed:
Move storage health_info logic into storage.rb.

- Refactor ManageNextcloudIntegration* jobs. Add specs.
- Add model specs.

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into impl/add-health-info-to-storages

ba1ash committed:
Merge branch 'dev' into impl/add-health-info-to-storages

ba1ash committed:
Preload project to avoid n+1 queries.

ba1ash committed:
Handle a case when health_status is nil.

apfohl commented on modules/storages/app/models/storages/storage.rb:
```diff
@@ -98,6 +88,24 @@ def self.shorten_provider_type(provider_type)
       end
     end
 
+    def mark_as_unhealthy(reason: nil)
+      update(health_status: 'unhealthy', health_changed_at: Time.now.utc, health_reason: reason)
+    end
+
+    def mark_as_healthy
+      update(health_status: 'healthy', health_changed_at: Time.now.utc, health_reason: nil)
+    end
+
+    def healthy?
+      mark_as_healthy if health_status.nil?
+      health_status == 'healthy'
+    end
+
+    def unhealthy?
+      mark_as_healthy if health_status.nil?
+      health_status == 'unhealthy'
+    end
+
```
I Suggest adding a default value to the health status in the database and simplify the logic here to side effect free methods.

apfohl replied:
I suggest `unhealthy` as the default and setting it to `health` upon completion of the setup.

apfohl commented on modules/storages/spec/models/storages/nextcloud_storage_spec.rb:
```diff
@@ -42,6 +42,74 @@
+      expect(storage).to be_healthy
+    end
+
+    it 'does not mark as healthy if health_status has been defined earlier' do
+      storage.mark_as_healthy
+      expect do
+        storage.unhealthy?
+      end.not_to(change(storage, :health_status).from('healthy'))
+      expect(storage).not_to be_unhealthy
+    end
```
With the changes above in the storage model these specs need to be reworked.

apfohl reviewed (changes requested):
Please create a default value for the health status as described in my comments.

ba1ash committed:
Set storages.health_info default on the database level.

ba1ash committed:
Remove jsonb health_info and use usual columns.

No need for having jsonb for now.

ba1ash committed:
Merge branch 'dev' into impl/add-health-info-to-storages

ba1ash committed:
Merge branch 'dev' into impl/add-health-info-to-storages

ba1ash committed:
Rename unconfigured to pending.

Kharonus reviewed (approved):
I have nothing to add to this change anymore. So I agree for a merge. If we find anything not fitting, we can still adapt the code. from maintenance perspective code looks good! 👍🏽
